src/home/homework-fscrypt.c

```diff
@@ -0,0 +1,644 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <linux/fs.h>
+#include <openssl/evp.h>
+#include <openssl/sha.h>
+#include <sys/ioctl.h>
+#include <sys/xattr.h>
+
+#include "errno-util.h"
+#include "fd-util.h"
+#include "hexdecoct.h"
+#include "homework-fscrypt.h"
+#include "homework-quota.h"
+#include "memory-util.h"
+#include "missing_keyctl.h"
+#include "missing_syscall.h"
+#include "mkdir.h"
+#include "nulstr-util.h"
+#include "openssl-util.h"
+#include "parse-util.h"
+#include "process-util.h"
+#include "random-util.h"
+#include "rm-rf.h"
+#include "stdio-util.h"
+#include "strv.h"
+#include "tmpfile-util.h"
+#include "user-util.h"
+#include "xattr-util.h"
+
+static int fscrypt_upload_volume_key(
+                const uint8_t key_descriptor[static FS_KEY_DESCRIPTOR_SIZE],
+                const void *volume_key,
+                size_t volume_key_size,
+                key_serial_t where) {
+
+        _cleanup_free_ char *hex = NULL;
+        const char *description;
+        struct fscrypt_key key;
+        key_serial_t serial;
+
+        assert(key_descriptor);
+        assert(volume_key);
+        assert(volume_key_size > 0);
+
+        if (volume_key_size > sizeof(key.raw))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Volume key too long.");
+
+        hex = hexmem(key_descriptor, FS_KEY_DESCRIPTOR_SIZE);
+        if (!hex)
+                return log_oom();
+
+        description = strjoina("fscrypt:", hex);
+
+        key = (struct fscrypt_key) {
+                .size = volume_key_size,
+        };
+        memcpy(key.raw, volume_key, volume_key_size);
+
+        /* Upload to the kernel */
+        serial = add_key("logon", description, &key, sizeof(key), where);
```

Comment:
Have you checked out the fscrypt changes in kernel v5.4?  There are new ioctls for adding/removing fscrypt keys which fix various problems with the old way, and a new encryption policy version which fixes some security limitations of the old version.  See https://www.kernel.org/doc/html/latest/filesystems/fscrypt.html

Since systemd-homed is new code, if you don't mind requiring kernel v5.4+ for fscrypt to be used with systemd-homed, you might want to consider supporting v2 encryption policies only.

Note that the other fscrypt maintainers and I aren't planning to add any new features to v1 encryption policies and the process-subscribed keyring based API for fscrypt.  They're maintained for backwards compatibility only.

It also looks like systemd-homed never "locks" the directory after it's been unlocked.  With the new FS_IOC_REMOVE_ENCRYPTION_KEY ioctl it would be easy to do

Reply:
Yupp, I had a look. We can certainly move to that, but so far my own development distro doesn't even have that kernel, and until that happens I am too lazy to implement it... ;-)

Reply:
Okay.  Just a heads up, with the old API people will run into issues like https://github.com/google/fscrypt/issues/116, https://github.com/google/fscrypt/issues/128, and https://bugs.launchpad.net/ubuntu/+source/fscrypt/+bug/1770715.  These are fixed by the new API.

Reply:
> Yupp, I had a look. We can certainly move to that, but so far my own development distro doesn't even have that kernel, and until that happens I am too lazy to implement it... ;-)
5.4.2 is in koji, but I suspect F31 rebase will be next week at the earliest, either 5.4.3 or 5.4.4, due to some regressions. Seems sanity promoting to only implement the new fscrypt API, clean break.